### PR TITLE
GPG Key Ownership verification with Signed Token

### LINKED ISSUE #14036: Support adding GPG signing sub-key
**Type** : Feature request

## Description
As @a1012112796 has mentioned in https://github.com/go-gitea/gitea/issues/14009#issuecomment-746558233 "Now gitea only allow add public gpg keys bounded with the user's email", it would be nice if Gitea also support adding GPG sub-key, just like mention in issue https://github.com/go-gitea/gitea/issues/14009.

### PR DESCRIPTION
**Pull request by zeripath**
Currently we will only allow a key to be matched to a user if it matches
an activated email address and will only verify commits if the committer email address
matches an activated email address in the key . This PR provides a different mechanism - if
the user provides a signature for automatically generated token (based
on the timestamp, user creation time, user ID, username and primary
email. This marks the key as verified and allows commits signed by this key, where the 
committer email address matches any activated email address for the user to be verified.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### Matching Identity - Unverified
unverified-matched
matched-popup
(Pop up now reads: The embedded identities in this key match the following activated email addresses for this user and commits matching these email addresses can be verified with this key.)

## Attempt Token Verification
matched-do-verify

## Failed Verifcation
failed-signature

## Matched and Verified
verified-matched
verified-popup

## No Matching Identity
not-matched-message
provide-signature

## Verified Not Matched
verified-not-matched

Fixes #14036

### CODE REVIEW

**Comment by zeripath** [comment]
So one question here is whether the round to nearest 5 minutes is reasonable or whether we should make them resubmit the token back. 

Round to nearest 5 minutes gives a potential issue whereby a token could be out of date by the time it's shown to the user.

**Comment by zeripath** [comment]
Should fix #14036

**Comment by 6543** [comment]
@zeripath can you resolve conflicts?

**Comment by zeripath** [comment]
I'm gonna move this to 1.15 as it doesn't appear to have been tested by anyone, and in particular any of the people who were interested in the feature.

File: integrations/api_gpg_keys_test.go [hunk 1]
```diff
@@ -77,6 +77,9 @@ func TestGPGKeys(t *testing.T) {
 		DecodeJSON(t, resp, &keys)
```
Comment (6543) [inline integrations/api_gpg_keys_test.go:76]: Assert len keys = 2

**Comment by 6543** [comment]
I'm still not sure if it is good, to add a key without checking if user also controls related email :/

with this I can create a key for email another user controle & let the commit show valide

**Comment by zeripath** [comment]
> I'm still not sure if it is good, to add a key without checking if user also controls related email :/

The emails still have to be activated.

> with this I can create a key for email another user controle & let the commit show valide

You can do that already if you allow auto vaidated email addresses by just adding another email address to your gpg key.

File: routers/api/v1/user/gpg_key.go [hunk 2]
```diff
@@ -119,14 +123,84 @@ func GetGPGKey(ctx *context.APIContext) {
 
 // CreateUserGPGKey creates new GPG key to given user by ID.
 func CreateUserGPGKey(ctx *context.APIContext, form api.CreateGPGKeyOption, uid int64) {
-	keys, err := models.AddGPGKey(uid, form.ArmoredKey)
+	token := base.EncodeSha256(time.Now().Truncate(1*time.Minute).Add(1*time.Minute).Format(time.RFC1123Z) + ":" + ctx.User.CreatedUnix.FormatLong() + ":" + ctx.User.Name + ":" + ctx.User.Email + ":" + strconv.FormatInt(ctx.User.ID, 10))
```
Comment (6543) [inline routers/api/v1/user/gpg_key.go]: look's like we should put the token generation into it's won func?

to make code cleaner and dedub it

Comment (zeripath) [reply routers/api/v1/user/gpg_key.go]: Done

**Comment by zeripath** [comment]
I suspect 18df555 might break tests btw @6543

**Review by 6543 — APPROVED** [review]

File: models/gpg_key_verify.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	if err := sess.Commit(); err != nil {
+		return "", err
+	}
+
+	return key.KeyID, nil
+}
+
+// VerificationToken returns token for the user that will be valid in minutes minutes time
```
Comment (6543) [inline models/gpg_key_verify.go]:
```suggestion
// VerificationToken returns token for the user that will be valid in minutes (time)
```

**Comment by 6543** [comment]
@zeripath #12155 did create confilcts - would be nice if you have time to resolve

File: models/gpg_key_add.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"strings"
+
+	"code.gitea.io/gitea/modules/log"
+	"github.com/keybase/go-crypto/openpgp"
```
Comment (techknowlogick) [inline models/gpg_key_add.go:12]:
```suggestion

	"github.com/keybase/go-crypto/openpgp"
```

File: models/gpg_key_commit_verification.go [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+import (
+	"container/list"
+	"fmt"
+	"hash"
+	"strings"
+
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"github.com/keybase/go-crypto/openpgp/packet"
```
Comment (techknowlogick) [inline models/gpg_key_commit_verification.go:17]:
```suggestion

	"github.com/keybase/go-crypto/openpgp/packet"
```

**Review by techknowlogick — APPROVED** [review]
Tested this PR works, code looks good, fixed a few minor nits. All good to merge once conflicts are resolved.

**Comment by axifive** [comment]
fixed #15263

**Pull request merged**